engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorImpl.java

```diff
@@ -156,7 +158,17 @@ public ValidatorImpl(ConstraintValidatorFactory constraintValidatorFactory,
 		this.validationOrderGenerator = validationOrderGenerator;
 		this.traversableResolver = traversableResolver;
 		this.validatorScopedContext = new ValidatorScopedContext( messageInterpolator, parameterNameProvider, clockProvider, temporalValidationTolerance, scriptEvaluatorFactory, failFast, traversableResolverResultCacheEnabled );
-		this.constraintValidatorInitializationContext = new HibernateConstraintValidatorInitializationContextImpl( validatorScopedContext.getScriptEvaluatorFactory(), validatorScopedContext.getClockProvider(), validatorScopedContext.getTemporalValidationTolerance() );
+		this.scriptEvaluatorFactory = validatorScopedContext.getScriptEvaluatorFactory();
+		this.constraintValidatorInitializationContext = new HibernateConstraintValidatorInitializationContextImpl( validatorScopedContext.getScriptEvaluatorFactory(), validatorScopedContext.getClockProvider(), validatorScopedContext.getTemporalValidationTolerance(), null );
+	}
+
+	public ValidatorImpl withDynamicConstraintValidatorInitializationPayload(Object dynamicPayload) {
+		this.constraintValidatorInitializationContext = new HibernateConstraintValidatorInitializationContextImpl(
+			this.scriptEvaluatorFactory,
```

Comment:
Instead of saving a reference to the `ScriptEvaluatorFactory` here inside the `ValidatorImpl` class we could also add a `getScriptEvaluatorFactory()` to the `(Hibernate)ConstraintValidatorInitializationContext(Impl)` - if it would be ok to expose that `ScriptEvaluatorFactory` from that context.
We could then replace this line here with `this.constraintValidatorInitializationContext.getScriptEvaluatorFactory()` instead. I leave it up to you which approach to use.

Reply:
We sure don't want to expose it (see the `clear()` method for instance).

What you could do instead is creating another class that wraps the original `HibernateConstraintValidatorInitializationContextImpl` and adds the payload. The original `HibernateConstraintValidatorInitializationContextImpl` would simply return null for the payload.